src/include/ipc/info.h

```diff
@@ -110,8 +110,8 @@ struct sof_ipc_window {
 	struct sof_ipc_window_elem window[];
 } __attribute__((packed));
 
-/* extended data, compiler version */
```

Comment:
Looks like this patch is trying to solve a linker issue ?

Reply:
Exactly

Reply:
Linker script ?

Reply:
Are you suggesting to resolve problem using linker script? If yes, then there was also error during compilation wrapped struct with flex array inside: `Error: attempt to move .org backwards`

Reply:
This is correct fix. I can keep `struct sof_ipc_cc_version` untouched and choose `.fw_metadata` structs declaration order to make it works with this compiler disability and it will looks lovely in PR - but it will just hide the problem. The problem come back when next person will try to add new struct to `.fw_metadata`. In this solution with `sof_ipc_cc_version_base`.

Just try to add following code to source and run compilation:

```
struct flex_struct {
	uint32_t header;
	uint32_t size;
	char flex[];
};

struct wrapped_flex {
	uint32_t header;
	uint32_t size;
	struct flex_struct fs;
};

#define TEXT "{'a', 'b', 'c'}}\0"
#define EXTRA_SIZE (-1)

const struct flex_struct instance11 __section(".fw_ready_metadata") = {
	.size = sizeof(struct flex_struct) + sizeof(TEXT),
	.header = 'a',
	.flex = TEXT,
};

const struct flex_struct instance12 __section(".fw_ready_metadata") = {
	.size = sizeof(struct flex_struct) + sizeof(TEXT),
	.header = 'b',
	.flex = TEXT,
};

/* can be more flex_struct in .fw_ready_metadata, no problem */
/* wrapped_flex can be one or none (memory layout dependant) */

const struct wrapped_flex instance21 __section(".fw_ready_metadata") = {
	.size = sizeof(struct wrapped_flex) + sizeof(TEXT),
	.header = 333,
	.fs = {
		.header = 'c',
		.flex = TEXT,
	},
};

/* Folowing declaration will rise compilation (not linker) error
const struct wrapped_flex instance22 __section(".fw_ready_metadata") = {
	.size = sizeof(struct wrapped_flex) + sizeof(TEXT),
	.header = 333,
	.fs = {
		.header = 'd',
		.flex = TEXT,
	},
};
*/
```

Reply:
Is the data too big for the section ? i.e. compiler/assembler attempts to move it (xtensa has some restrictions on placements)

Reply:
I don't think so, there's no problem with appending constant size structures (or structs with flex array defined explicit, not encapsulated inside another struct) and it is compilation error, not linking - so at compilation time sections haven't any size restrictions.

Reply:
Can you paste the full errors.

Reply:
```
[  2%] Building C object src/kernel/CMakeFiles/ext_manifest.dir/ext_manifest.c.o
cd /home/admin/ktrzcinskix/sof/build/src/kernel && /home/admin/Documents/Tools/Xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc -DCC_OPTIMIZE_FLAGS=\"O2\" -DXCC_TOOLS_VERSION=\"RG-2017.8-linux\" -DRELATIVE_FILE=\"src/kernel/ext_manifest.c\" -I/home/admin/ktrzcinskix/sof/src/platform/intel/cavs/include -I/home/admin/ktrzcinskix/sof/src/arch/xtensa/include -I/home/admin/ktrzcinskix/sof/src/arch/xtensa/xtos -I/home/admin/ktrzcinskix/sof/src/platform/cannonlake/include -I/home/admin/Documents/Tools/Xtensa/RG-2017.8-linux/X6H3CNL_2017_8/xtensa-elf/arch/include -I/home/admin/Documents/Tools/Xtensa/RG-2017.8-linux/X6H3CNL_2017_8/xtensa-elf/include -I/home/admin/ktrzcinskix/sof/src/include -I/home/admin/ktrzcinskix/sof/build/generated/include  -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -o CMakeFiles/ext_manifest.dir/ext_manifest.c.o   -c /home/admin/ktrzcinskix/sof/src/kernel/ext_manifest.c
/tmp/cc0spin#1d9701aa.UQdy3b: Assembler messages:
/tmp/cc0spin#1d9701aa.UQdy3b:77: Error: attempt to move .org backwards
/tmp/cc0spin#1d9701aa.UQdy3b:104: Error: attempt to move .org backwards
/tmp/cc0spin#1d9701aa.UQdy3b:137: Error: attempt to move .org backwards
xt-xcc ERROR:  /home/admin/Documents/Tools/Xtensa/RG-2017.8-linux/XtensaTools/bin/xt-as returned non-zero status 1
src/kernel/CMakeFiles/ext_manifest.dir/build.make:62: recipe for target 'src/kernel/CMakeFiles/ext_manifest.dir/ext_manifest.c.o' failed
make[2]: *** [src/kernel/CMakeFiles/ext_manifest.dir/ext_manifest.c.o] Error 2
```

Reply:
@ktrzcinx see here https://stackoverflow.com/questions/58537775/assembler-messages-error-attempt-to-move-org-backwards
It does seem like your structure is too big for the section and causes the assembler to fail on code it cant move. Can you check locally with @mmaka1 and @tlauda.

Reply:
>causes the assembler to fail on code it cant move

Introduced structures are unused from source code

Reply:
ok, but I don't want this worked around, only to hit someone else next week. Its also possible a new section definition or some other data/text location changes are the root cause.

Reply:
>  Its also possible a new section definition or some other data/text location changes are the root cause.

I've tried different sections, also using default one (without __section attribute), with the same result. Moreover before linking process, there is no a lot of place to make error in section definitions.
After changing toolchain to GCC 8.1.0 I get rid of this error, so I claim it's a XCC bug (or bug in old GCC toolchain but it's irrelevant).
As I know there are 5 solutions:
1) change `sof_ipc_cc_version.desc` field type to fixed size array - regression
2) copy-paste  `sof_ipc_cc_version` definition content to `ext_man_cc_version` - lack of kernel code reusability
3) extracting `sof_ipc_cc_version_base` from `sof_ipc_cc_version`
4) permanent remove cc_version from mailbox as it's only debug information (not needed to keep fw normaly working with kernel), keep it only in ext_man
5) remove `ext_man_elem_header` and use `sof_ipc_ext_data_hdr` instead

So as third solution is going to be rejected, then I'm going to implement 5'th

Reply:
Or wait for https://github.com/thesofproject/sof/pull/2522 as it is also solution for this problem

Reply:
lets fix this part with #2522

Reply:
changed